oxidation/crates/parsec_api_protocol/src/handshake.rs

```diff
@@ -0,0 +1,659 @@
+#[serde(rename_all = "snake_case")]
+pub enum HandshakeResult {
+    BadAdminToken,
+    BadIdentity,
+    BadProtocol,
+    Ok,
+    OrganizationExpired,
+    RevokedDevice,
+    RvkMismatch,
+}
+
+impl_dumps_loads!(Answer);
+
+#[serde_as]
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+#[serde(tag = "handshake", rename_all = "snake_case")]
+pub enum Handshake {
+    Challenge {
+        #[serde_as(as = "Bytes")]
+        challenge: Vec<u8>,
+        supported_api_versions: Vec<ApiVersion>,
+        // Those fields have been added to API version 2.4
+        // They are provided to the client in order to allow them to detect whether
+        // their system clock is out of sync and let them close the connection.
+        // They will be missing for older backend so they cannot be strictly required.
+        // TODO: This backward compatibility should be removed once Parsec < 2.4 support is dropped
+        #[serde(default, deserialize_with = "maybe_field::deserialize_some")]
+        ballpark_client_early_offset: Option<f64>,
+        #[serde(default, deserialize_with = "maybe_field::deserialize_some")]
+        ballpark_client_late_offset: Option<f64>,
+        #[serde(default, deserialize_with = "maybe_field::deserialize_some")]
+        backend_timestamp: Option<DateTime>,
+    },
+    Answer(Box<Answer>),
+    Result {
+        result: HandshakeResult,
+        #[serde(default, deserialize_with = "maybe_field::deserialize_some")]
+        help: Option<String>,
+    },
+}
+
+impl_dumps_loads!(Handshake);
+
+#[derive(Debug)]
+pub struct ServerHandshakeStalled {
+    pub challenge_size: usize,
+    pub supported_api_version: Vec<ApiVersion>,
+}
+
+impl Default for ServerHandshakeStalled {
+    fn default() -> Self {
+        Self {
+            challenge_size: 48,
+            supported_api_version: vec![API_V2_VERSION, API_V1_VERSION],
+        }
+    }
+}
+
+impl ServerHandshakeStalled {
+    pub fn new(challenge_size: usize) -> Self {
+        Self {
+            challenge_size,
+            supported_api_version: vec![API_V2_VERSION, API_V1_VERSION],
+        }
+    }
+    pub fn build_challenge_req(self) -> Result<ServerHandshakeChallenge, HandshakeError> {
+        let mut challenge = vec![0; self.challenge_size];
+        thread_rng().fill(&mut challenge[..]);
+
+        let raw = Handshake::Challenge {
+            challenge: challenge.clone(),
+            supported_api_versions: self.supported_api_version.clone(),
+            ballpark_client_early_offset: Some(BALLPARK_CLIENT_EARLY_OFFSET),
+            ballpark_client_late_offset: Some(BALLPARK_CLIENT_LATE_OFFSET),
+            backend_timestamp: Some(DateTime::now()),
+        }
+        .dumps()?;
+
+        Ok(ServerHandshakeChallenge {
+            supported_api_version: self.supported_api_version,
+            challenge,
+            raw,
+        })
+    }
+    #[cfg(feature = "test")]
+    pub fn build_challenge_req_with_challenge(
```

Comment:
would be better to have `build_challenge_req` and `build_challenge_req_with_challenge` rely on the same internal `_build_challenge_req_with_challenge` function instead of duplicating their internals (this is to ensure there won't be drifting between the two in later refactoring)

Reply:
I agree :+1;

Reply:
Done :+1: